axlearn/common/attention.py

```diff
@@ -168,10 +169,160 @@ class ForwardMode(enum.Enum):
+    """Represents key/value projections.
+
+    Fields:
+        k_proj: [batch, source_length, num_kv_heads, per_head_dim], Projected key tensor.
+        v_proj: [batch, source_length, num_kv_heads, per_head_dim], Projected value tensor.
+        key_positions: [batch, source_length], Positions of the keys in the batch.
+    """
 
     k_proj: Tensor
     v_proj: Tensor
+    key_positions: Tensor
+
+
+class BaseKVCache(BaseLayer):
+    """Abstract base class for KV cache."""
+
+    @config_class
+    class Config(BaseLayer.Config):
+        """Configures BaseKVCache."""
+
+        # Autoregressive cache dtype. Should match the step dtype.
+        # Needs to match the forward dtype for Repeated layers. If None, infer as BaseLayer.dtype().
+        cache_dtype: Optional[jnp.dtype] = None
+
+    class Output(KVState):
+        pass
+
+    class Shape(NamedTuple):
+        """Shape of BaseKVCache.
+
+        Fields:
+            batch_size: int, batch size.
+            kv_len: int, KV Cache size.
+            num_kv_heads: int, the number of KV heads.
+            per_head_dim: int, the dimension per head.
+        """
+
+        batch_size: int
+        kv_len: int
+        num_kv_heads: int
+        per_head_dim: int
+
+    def _cache_dtype(self, dtype: jnp.dtype):
+        # Default to activation dtype for initialization if cache_dtype is None.
+        dtype = self.config.cache_dtype or dtype
+        assert dtype is not None
+        return dtype
+
+    def init_states(self, shape: Shape, *, dtype: jnp.dtype) -> Nested[Tensor]:
+        """Initializes KV cache.
+
+        Args:
+            shape: Shape, [batch, kv_len, num_kv_heads, per_head_dim].
+            dtype: dtype for KV Cache.
+
+        Returns:
+            init_state: A `Nested[Tensor]` object containing KV cache such as key and value.
+        """
+        raise NotImplementedError(type(self))
+
+    def extend_step(
+        self,
+        cached_states: Nested[Tensor],
+        *,
+        k_proj: Tensor,
+        v_proj: Tensor,
+        proj_positions: Tensor,
+        paddings: Optional[Tensor] = None,
+    ) -> tuple[Nested[Tensor], Output]:
+        """Updates the KV cache per extend step.
+
+        The input k_proj/v_proj are generated by `i_proj.forward()` from the input tokens.
+        The output k_proj/v_proj will have the `kv_cache` concatenated and will be used
+        in the current attention operation.
+
+        Args:
+            cached_states: A `Nested[Tensor]` object containing KV cache such as key and value.
+            k_proj: [batch, source_length, num_kv_heads, per_head_dim], Projected key tensor.
+            v_proj: [batch, source_length, num_kv_heads, per_head_dim], Projected value tensor.
+            proj_positions: [batch, source_length], Positions of the k_proj, int32.
+            paddings: [batch, source_length], An optional paddings of the k_proj, k_proj dtype.
```

Comment:
nit: Could you please add more details on the semantics of paddings?

Reply:
it's changed to live_step_len and has long docstring.